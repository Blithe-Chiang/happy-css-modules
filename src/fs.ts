import * as fs from 'fs';


export function readFile(path: fs.PathLike): Promise<string> {
    return new Promise<string>((resolve, reject) => {
        fs.readFile(path, 'utf-8', (err, data) => {
            if (err !== null) {
                reject(err);
                return;
            }

            resolve(data);
        });
    });
}

export function writeFile(path: fs.PathLike, data: string): Promise<void> {
    return new Promise<void>((resolve, reject) => {
        fs.writeFile(path, data, 'utf-8', (err) => {
            if (err !== null) {
                reject(err);
                return;
            }

            resolve();
        });
    });
}
