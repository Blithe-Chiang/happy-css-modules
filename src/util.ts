/**
 * The SystemError type of Node.js.
 * @see https://nodejs.org/api/errors.html#class-systemerror
 */
export interface SystemError {
  code: string;
}

export function isSystemError(value: unknown): value is SystemError {
  return (
    isObject(value) &&
    hasProp(value, 'constructor') &&
    isObject(value.constructor) &&
    hasProp(value.constructor, 'name') &&
    value.constructor.name === 'Error' &&
    hasProp(value, 'code') &&
    typeof value.code === 'string'
  );
}

export function isObject(value: unknown): value is object {
  return (typeof value === 'object' || typeof value === 'function') && value !== null;
}

export function hasProp<T extends string>(obj: object, prop: T): obj is { [key in T]: unknown } {
  return prop in obj;
}

export function unique<T>(array: T[]): T[] {
  return [...new Set(array)];
}

export function uniqueBy<T, U>(arr: T[], fn: (el: T) => U): T[] {
  const result: T[] = [];
  const keys = new Set<U>();
  for (const el of arr) {
    const key = fn(el);
    if (!keys.has(key)) {
      keys.add(key);
      result.push(el);
    }
  }
  return result;
}
