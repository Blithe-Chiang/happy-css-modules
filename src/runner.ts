import { resolve } from 'path';
import * as process from 'process';
import * as util from 'util';
import chalk from 'chalk';
import * as chokidar from 'chokidar';
import _glob from 'glob';
import { emitGeneratedFiles } from './emitter/index.js';
import { Loader, Transformer } from './loader.js';

const glob = util.promisify(_glob);

export type Watcher = {
  close: () => Promise<void>;
};

export type LocalsConvention = 'camelCase' | 'camelCaseOnly' | 'dashes' | 'dashesOnly' | undefined;

export interface RunnerOptions {
  pattern: string;
  outDir?: string;
  watch?: boolean;
  localsConvention?: LocalsConvention;
  namedExport?: boolean;
  declarationMap?: boolean;
  transform?: Transformer;
  /**
   * Silent output. Do not show "files written" messages.
   * @default false
   */
  silent?: boolean;
  /** Working directory path. */
  cwd?: string;
}

type OverrideProp<T, K extends keyof T, V extends T[K]> = Omit<T, K> & { [P in K]: V };

/**
 * Run typed-css-module.
 * @param options Runner options.
 * @returns Returns `Promise<Watcher>` if `options.watch` is `true`, `Promise<void>` if `false`.
 */
export async function run(options: OverrideProp<RunnerOptions, 'watch', true>): Promise<Watcher>;
export async function run(options: RunnerOptions): Promise<void>;
export async function run(options: RunnerOptions): Promise<Watcher | void> {
  const loader = new Loader(options.transform);
  const distOptions = options.outDir
    ? {
        rootDir: process.cwd(), // TODO: support `--rootDir` option
        outDir: options.outDir,
      }
    : undefined;

  async function processFile(filePath: string) {
    try {
      const result = await loader.load(filePath);
      await emitGeneratedFiles({
        filePath,
        tokens: result.tokens,
        distOptions,
        emitDeclarationMap: options.declarationMap,
        dtsFormatOptions: {
          localsConvention: options.localsConvention,
          namedExport: options.namedExport,
        },
        silent: options.silent ?? false,
        cwd: options.cwd ?? process.cwd(),
      });
    } catch (error) {
      // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
      console.error(chalk.red('[Error] ' + error));
    }
  }

  if (options.watch) {
    if (!options.silent) console.log('Watch ' + options.pattern + '...');
    const watcher = chokidar.watch([options.pattern.replace(/\\/g, '/')]);
    watcher.on('all', (eventName, filePath) => {
      if (eventName === 'add' || eventName === 'change') {
        void processFile(filePath);
      }
    });
    return { close: async () => watcher.close() };
  } else {
    const filePaths = (await glob(options.pattern, { dot: true }))
      // convert relative path to absolute path
      .map((file) => resolve(file));

    // TODO: Use `@file-cache/core` to process only files that have changed
    for (const filePath of filePaths) {
      await processFile(filePath);
    }
  }
}
