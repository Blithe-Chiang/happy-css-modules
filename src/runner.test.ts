import { readFile, writeFile } from 'fs/promises';
import { jest } from '@jest/globals';
import chalk from 'chalk';
import AggregateError from 'es-aggregate-error';
import { run } from './runner.js';
import { createFixtures, exists, getFixturePath, transformer, waitForAsyncTask } from './test/util.js';

// eslint-disable-next-line @typescript-eslint/no-empty-function
const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

const defaultOptions = {
  pattern: 'test/**/*.{css,scss}',
  declarationMap: true,
  silent: true,
  cwd: getFixturePath('/'),
};

test('generates .d.ts and .d.ts.map', async () => {
  createFixtures({
    '/test/1.css': '.a {}',
    '/test/2.css': '.b {}',
  });
  await run({ ...defaultOptions });
  expect(await readFile(getFixturePath('/test/1.css.d.ts'), 'utf8')).toMatchSnapshot();
  expect(await readFile(getFixturePath('/test/1.css.d.ts.map'), 'utf8')).toMatchSnapshot();
  expect(await readFile(getFixturePath('/test/2.css.d.ts'), 'utf8')).toMatchSnapshot();
  expect(await readFile(getFixturePath('/test/2.css.d.ts.map'), 'utf8')).toMatchSnapshot();
});

test.todo('changes dts format with camelCase and namedExport options');
test('does not emit declaration map if declarationMap is false', async () => {
  createFixtures({
    '/test/1.css': '.a {}',
  });
  await run({ ...defaultOptions, declarationMap: false });
  await expect(readFile(getFixturePath('/test/1.css.d.ts'), 'utf8')).resolves.not.toThrow();
  await expect(readFile(getFixturePath('/test/1.css.d.ts.map'), 'utf8')).rejects.toThrow(/ENOENT/);
});
test('supports transformer', async () => {
  createFixtures({
    '/test/1.scss': `.a { dummy: ''; }`,
  });
  await run({ ...defaultOptions, transformer });
  expect(await readFile(getFixturePath('/test/1.scss.d.ts'), 'utf8')).toMatchSnapshot();
  expect(await readFile(getFixturePath('/test/1.scss.d.ts.map'), 'utf8')).toMatchSnapshot();
});
test('watches for changes in files', async () => {
  createFixtures({
    '/test': {}, // empty directory
  });
  const watcher = await run({ ...defaultOptions, watch: true });

  await writeFile(getFixturePath('/test/1.css'), '.a-1 {}');
  await waitForAsyncTask(500); // Wait until the file is written
  expect(await readFile(getFixturePath('/test/1.css.d.ts'), 'utf8')).toMatch(/a-1/);

  await writeFile(getFixturePath('/test/1.css'), '.a-2 {}');
  await waitForAsyncTask(500); // Wait until the file is written
  expect(await readFile(getFixturePath('/test/1.css.d.ts'), 'utf8')).toMatch(/a-2/);

  await watcher.close();
});
test('returns an error if the file fails to process in non-watch mode', async () => {
  createFixtures({
    '/test/1.css': '.a {}',
    '/test/2.css': 'INVALID SYNTAX',
    '/test/3.css': 'INVALID SYNTAX',
  });
  const maybeError = await run({ ...defaultOptions, watch: false }).catch((e) => e);

  // The errors are aggregated into AggregateError.
  expect(maybeError).toBeInstanceOf(AggregateError);
  const error = maybeError as AggregateError;
  expect(error.message).toMatchInlineSnapshot(`"Failed to process files"`);
  expect(error.errors).toHaveLength(2);
  expect(error.errors[0]).toMatchInlineSnapshot(`<fixtures>/test/2.css:1:1: Unknown word`);
  expect(error.errors[1]).toMatchInlineSnapshot(`<fixtures>/test/3.css:1:1: Unknown word`);

  // The error is logged to console.error.
  expect(consoleErrorSpy).toHaveBeenCalledTimes(2);
  // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
  expect(consoleErrorSpy).toHaveBeenNthCalledWith(1, chalk.red('[Error] ' + error.errors[0]));
  // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
  expect(consoleErrorSpy).toHaveBeenNthCalledWith(2, chalk.red('[Error] ' + error.errors[1]));

  // The valid files are emitted.
  expect(await exists(getFixturePath('/test/1.css.d.ts'))).toBe(true);
  expect(await exists(getFixturePath('/test/1.css.d.ts.map'))).toBe(true);
});
