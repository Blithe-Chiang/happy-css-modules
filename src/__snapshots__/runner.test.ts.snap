// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates .d.ts and .d.ts.map 1`] = `
"declare const styles: {
  readonly "a": string;
};
export = styles;
//# sourceMappingURL=1.css.d.ts.map
"
`;

exports[`generates .d.ts and .d.ts.map 2`] = `"{"version":3,"sources":["1.css"],"names":["a"],"mappings":"AAAA;AAAA,E,SAAAA,G,SAAA;AAAA;AAAA","file":"1.css.d.ts","sourceRoot":""}"`;

exports[`generates .d.ts and .d.ts.map 3`] = `
"declare const styles: {
  readonly "b": string;
};
export = styles;
//# sourceMappingURL=2.css.d.ts.map
"
`;

exports[`generates .d.ts and .d.ts.map 4`] = `"{"version":3,"sources":["2.css"],"names":["b"],"mappings":"AAAA;AAAA,E,SAAAA,G,SAAA;AAAA;AAAA","file":"2.css.d.ts","sourceRoot":""}"`;

exports[`supports transformer 1`] = `
"declare const styles: {
  readonly "a": string;
};
export = styles;
//# sourceMappingURL=1.scss.d.ts.map
"
`;

exports[`supports transformer 2`] = `"{"version":3,"sources":["1.scss"],"names":["a"],"mappings":"AAAA;AAAA,E,SAAAA,G,SAAA;AAAA;AAAA","file":"1.scss.d.ts","sourceRoot":""}"`;
